DEV: Chat service objects

martin-brennan opened the pull request:
This is a combined work of Martin Brennan, Loïc Guitaut, and Joffrey Jaffeux.

---

This commit implements a base service object when working in chat. The documentation is available at https://discourse.github.io/discourse/chat/backend/Chat/Service.html

Generating documentation has been made as part of this commit with a bigger goal in mind of generally making it easier to dive into the chat project.

Working with services generally involves 3 parts:

- The service object itself, which is a series of steps where few of them are specialized (model, transaction, policy)

```ruby
class UpdateAge
  include Chat::Service::Base

  model :user, :fetch_user
  policy :can_see_user
  contract
  step :update_age

  class Contract
    attribute :age, :integer
  end

  def fetch_user(user_id:, **)
    User.find_by(id: user_id)
  end

  def can_see_user(guardian:, **)
    guardian.can_see_user(user)
  end

  def update_age(age:, **)
    user.update!(age: age)
  end
end
```

- The `with_service` controller helper, handling success and failure of the service within a service and making easy to return proper response to it from the controller

```ruby
def update
  with_service(UpdateAge) do
    on_success { render_serialized(result.user, BasicUserSerializer, root: "user") }
  end
end
```

- Rspec matchers and steps inspector, improving the dev experience while creating specs for a service

```ruby
RSpec.describe(UpdateAge) do
  subject(:result) do
    described_class.call(guardian: guardian, user_id: user.id, age: age)
  end

  fab!(:user) { Fabricate(:user) }
  fab!(:current_user) { Fabricate(:admin) }

  let(:guardian) { Guardian.new(current_user) }
  let(:age) { 1 }

   it { expect(user.reload.age).to eq(age) }
end
```

Note in case of unexpected failure in your spec, the output will give all the relevant information:

```
  1) UpdateAge when no channel_id is given is expected to fail to find a model named 'user'
     Failure/Error: it { is_expected.to fail_to_find_a_model(:user) }

       Expected model 'foo' (key: 'result.model.user') was not found in the result object.

       [1/4] [model] 'user' ❌
       [2/4] [policy] 'can_see_user'
       [3/4] [contract] 'default'
       [4/4] [step] 'update_age'

       /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/update_age.rb:32:in `fetch_user': missing keyword: :user_id (ArgumentError)
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/base.rb:202:in `instance_exec'
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/base.rb:202:in `call'
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/base.rb:219:in `call'
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/base.rb:417:in `block in run!'
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/base.rb:417:in `each'
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/base.rb:417:in `run!'
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/base.rb:411:in `run'
       	from <internal:kernel>:90:in `tap'
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/app/services/base.rb:302:in `call'
       	from /Users/joffreyjaffeux/Code/pr-discourse/plugins/chat/spec/services/update_age_spec.rb:15:in `block (3 levels) in <main>'
```

martin-brennan commented on lib/service_result.rb:
```diff
@@ -0,0 +1,49 @@
+    @state == :success
+  end
+
+  def http_status
```
If we don't like this here as much, we could easily add something to `ApplicationController` that takes a `ServiceResult` object and spits this out (same for the checking for fail then rendering JSON error in the controller)

martin-brennan commented on plugins/chat/app/services/services.md:
```diff
@@ -0,0 +1,153 @@
+### Services
```
This document has my line of thinking, feel free to add comments

martin-brennan commented on lib/service_base.rb:
```diff
@@ -0,0 +1,112 @@
+    )
+  end
+
+  def execute_service_call
```
Wrapping everything in these `execute_service_call` calls lets us handle errors much more gracefully especially the uncaught ones, and makes sure we always are returning a `ServiceResult`

martin-brennan commented on plugins/chat/app/controllers/chat_controller.rb:
```diff
@@ -98,20 +98,22 @@ def create_message
+      return render_json_error(create_result.errors), status: create_result.http_status
+    end
 
+    # NOTE: All of this should probably be in the service too?? It's all to do with
```
You will need to look at the controller to see, but we are doing heaps of stuff here _after_ creating the message. Maybe that should be a separate service e.g. `NewChatMessageProcessor` or something like that...not sure

martin-brennan commented:
So far I am liking this, since it's much more standardized and feels cleaner to do things only the one way.

jjaffeux commented:
OK thanks will need some time to read and think about all of this :+1:

martin-brennan commented:
* Prefer raising normal errors (e.g. Discourse::InvalidAccess) for control flow, since they are handled in the controllers etc. If there are no errors raised assume success, and return a hash object if you need to give records back to the caller
* Do not force callers to always provide guardian
* Always use `.call` as a class method on the service object class
* Pass down `ActionController::Parameters`, then we can do validations on those in the service object including things like `.permit`. .to_json then create a new `ActionController::Parameters`. Probably just bake something on service base class to do this
* Do not double up `Chat` in class names, e.g. only `Chat::ChannelDestroyer`
* No need for things like `OpenStruct`

jjaffeux commented on plugins/chat/app/services/services.md:
```diff
@@ -0,0 +1,153 @@
+* Logging, both to STDOUT and to things like StaffActionLogger
+* Interacting with MessageBus
+
+When writing services, we also need to think about how we handle failure and
+success states, especially for the Controllers that are most often calling them
+(though of course services can and do call other services). With that in mind we
+need to handle each different function of services differently to make it easier
+for the developer.
```
I think there are two other places where it's going to be called a lot too:
- specs
- jobs

jjaffeux commented on plugins/chat/app/services/services.md:
```diff
@@ -0,0 +1,153 @@
+* Add all errors from ActiveRecord models when calling their own .validate
+  Methods
+
+Sometimes we need to stop the world when we find one of these errors, other
+times we can just collect them and show them all at the end. The base class for
+service objects should support both. We can define many ease-of-use methods in
```
Im mixed on this, errors that don't stop are not errors IMO. They are some kind of specific events in some services which could be treated in the result object:

```
if not_processed_count > 0
  context.audit_log << "#{not_processed_count} messages have not been processed"
end
```

And then have some logic in controller rendering if `audit_log` is not empty.

Failure should be a strong failure IMO.

jjaffeux commented on plugins/chat/app/services/services.md:
```diff
@@ -0,0 +1,153 @@
+work effort here, since we may end up checking permissions on something (a
+chat channel for example) multiple times.
+
+I don't think there is an easy solution to this, that doesn't involve some
+contortions around caching Guardian call signatures and their results for a
+specific user inside some sort of wrapper object. For now I think we can leave
+this, unless it's really causing a lot of performance woes.
```
I share the concern on this, we should closely follow how guardian will be used and seek for ways to improve it

jjaffeux commented on plugins/chat/app/services/services.md:
```diff
@@ -0,0 +1,153 @@
+specific user inside some sort of wrapper object. For now I think we can leave
+this, unless it's really causing a lot of performance woes.
+
+#### Enqueueing Jobs
+
+This can be done via the base class's `enqueue_job` and `enqueue_job_at` methods,
+rather than calling Jobs.enqueue directly.
+
+#### Logging
+
+As mentioned above, there are two types of logging:
+
+* STDOUT logging using Rails.logger and Discourse.warn_exception
+* UserHistory and StaffActionLogger database-level logs
+
+Both of these can be handled via the base class' `log_message`, `log_user_history`,
+and `log_staff_action` methods.
+
+#### MessageBus
+
+Ideally, any MessageBus interactions should be done via a dedicated MessageBus
+service, such as ChatMessagePublisher.
```
All of this shouldn't be part of the service utilities IMO. This is separate concerns.

jjaffeux commented on plugins/chat/app/services/services.md:
```diff
@@ -0,0 +1,153 @@
+Ideally, any MessageBus interactions should be done via a dedicated MessageBus
+service, such as ChatMessagePublisher.
+
+#### Data Payloads / Params
+
+Sometimes (e.g. for things like ChatMessageCreator) we require a lot of params
+for the service to work correctly. Instead of passing these through the
+initialize block, we should instead make a XPayload class within the service
+class and define with `attr_accessor` the params needed for the service. This
+can easily be done with `Class.new.tap { |x| x.param = y }` in the controller.
+
+Later, we may also want these payload classes to define their own validations of
+required parameters etc. if necessary, if we have a standard `.validate` method
+on all of them then we can pass results to the service class which can raise
+them to the base class.
```
Agreed this is a very important part of what we want to achieve/improve here 👌

jjaffeux commented:
@martin-brennan I pushed the results of your ideas, @Flink ideas and my ideas (and interactor :D). This is not finished, but I pushed it so we can talk of it.

Flink commented:
> One of the big missing piece IMO is to write some helper controller to handle failures.

We’d probably have something by Monday/Tuesday :grin:

jjaffeux commented:
@martin-brennan one of my goals with this work is also to provide better documentation:

https://discourse.github.io/discourse/chat/backend/Chat/Service/TrashChannel.html
https://discourse.github.io/discourse

Also made more tweaks to the service implementation.

martin-brennan commented on plugins/chat/assets/javascripts/discourse/services/chat-api.js:
```diff
@@ -70,17 +70,14 @@ export default class ChatApi extends Service {
   /**
    * Destroys a channel.
    * @param {number} channelId - The ID of the channel.
-   * @param {string} channelName - The name of the channel to be destroyed, used as confirmation.
```
I don't understand why we are removing this? It's an important part of the UI

jjaffeux replied:
Answered here: https://github.com/discourse/discourse/pull/19814#discussion_r1088748042

martin-brennan commented on plugins/chat/app/services/trash_channel.rb:
```diff
@@ -0,0 +1,70 @@
+
+      contract do
+        attribute :channel
+        validates :channel, presence: true
```
I noticed we are missing the `name_confirmation` here and in JS, not sure why this is removed? 

Also I was thinking of validation. We can validate the presence of it here, but how do we raise a validation error if the `name_confirmation` does not match the channel name? Do we just do an AM validation like so? (I am pretty sure we do):

```ruby
validate :name_confirmation_matches

def name_confirmation_matches
  errors.add(:name_confirmation, :invalid) if channel.title(guardian.user).downcase != name_confirmation
end

jjaffeux replied:
I don't think it needs server confirmation that's all :p Frontend only check is fine by me, it's footgun territory if you start to hack  your own js to be able to send a wrong name confirmation to the server.

But yes for the record you could do AM validation here, although guardian is not accessible here, I will wait to see if we actually need this at some point, because here technically you need a user not a guardian for example.

martin-brennan replied:
Fair enough, I guess you are right that it's a frontend only thing. I think I originally made it server-side as well because GitHub sends the confirmation in the delete repo payload. Not a hill I will die on though :)

martin-brennan commented on plugins/chat/app/services/base.rb:
```diff
@@ -0,0 +1,186 @@
+      module Helpers
+        def guardian(name, *args)
+          unless context[:guardian].public_send(name, *args)
+            context.fail!("guardian.failed" => name)
```
Actually even further, we could just add methods like this to the `Context` to make things nicer:

```ruby
def guardian_failed?
  guardian_failed_error.present?
end

def guardian_failed_error
  result[:guardian_failed]
end

def contract_failed?
  result[:contract_failed]
end
```

We don't want callers to have to dig into things IMO

jjaffeux replied:
So generally I agree with the `:"xxx"` being ugly and slightly odd. One thing to have in mind is that generally speaking we probably don't want people to interact much with it directly. eg: in controllers we will give a helper to handle all this logic for you.

This being said I wouldn't be against `:guardian_failed`, what do you think @Flink ?

martin-brennan replied:
Yeah I definitely think the controller can handle this generally, though we do have to keep in mind the Job -> Service and Service -> Service calls, which may need to inspect this manually. IMO the wrapper methods are a nicer way of accessing this since you don't need to know the hash key etc., but will see what Loic thinks too :+1:

Flink replied:
Yes, having a nice wrapper in controllers is probably what we want/need. This will provide a nice interface to work with. And in the end, a controller should not bother with too many logic details but just know how to handle HTTP stuff based on the service outcome.

Regarding the renaming, why not, the string keys were heavily inspired by Trailblazer. (where you can have things like `contract.default.result`, `policy.custom_name.result`, etc.)

martin-brennan reviewed (commented):
I like a lot what you've laid down here, I think it's a good evolution of what I had originally -- adding `ActiveModel`, taking ideas from `Interactor`, and generally making the base object a bit leaner. The documentation and spec additions are excellent too. I like that there are clear ways to write the services now, and there's only one way to do it really.

For the controller helper, I've pushed up an idea of how the helper might work, feel free to refine or trash. I don't think we need to manually handle many cases there, since we only have `contract.failed` (which we can render errors into `failed_json` for) and `guardian.failed` (which we can just raise `Discourse::InvalidAccess` for) to contend with. I've also fixed a couple of minor errors in the service base with private methods.

I was also confused about `name_confirmation` being left out for trash, which I've left some comments about. Another question I have is -- why put all these nice things like `plugins/chat/app/services/base.rb` and `plugins/chat/spec/support/chat_service_matcher.rb` into the chat namespace instead of just putting into core where it is generally useful? I know I would very much like to use these ideas for core services too. I guess we could move them later, but IMO it's a shame to just restrict it to chat.

I am going to create a `UpdateChannel` service today to try out what we've got.

Minor nitpick, for some reason https://discourse.github.io/discourse/chat/backend/index.html has things like `registerChatComposerButton`, shouldn't that be in the Frontend section?

martin-brennan commented:
@jjaffeux left some comments here, and also @Flink feel free to read and provide feedback as well :+1: I think we are getting quite close here, I am really eager to get this merged in so I can use it for upcoming work.

martin-brennan commented:
@jjaffeux I also just added an `UpdateChannel` service in this commit https://github.com/discourse/discourse/pull/19814/commits/24691b3cfc8a1aec04e543f749c228075ecfffc7 . It was quite easy, though I did come across a new problem which we needed to consider. That is the issue where, for controllers at least, not all the parameters can be passed in, in the case of the Update endpoint you can pass in the channel `:name` but not the `:description`, and we don't want to change `:description` to null just from the abscence of a parameter.

I came up with `change_attribute_defaults` for this. It takes in the original parameters passed to the contract and runs when the contract is created, we can use this to inspect the original params and do things like .e.g set the `name` attribute to `channel.name` if `params.key?(:name)` is false. Happy to discuss this and maybe come up with a better way of doing it / better name.

The other issue is rate limiting -- for the channels controller, based on the value of a parameter we perform `auto_join_limiter`. Do we want to make it so the controller can easily access the resolved service object contract data? Or have a way to allow services to handle rate limits (not sure if this is a good idea since it's mostly a controller thing ATM)? Or something else, maybe just leave it how it is in the controller? Not sure.

jjaffeux commented on plugins/chat/spec/support/chat_service_matcher.rb:
```diff
@@ -0,0 +1,25 @@
+RSpec::Matchers.define :succeed do
+  match { |context| context.success? === true }
+
+  failure_message { |context| CONTRACT_FAILED_ERROR.call(context) if context[:"contract.failed"] }
```
Will answer in the other.

jjaffeux commented on plugins/chat/spec/support/chat_service_matcher.rb:
```diff
@@ -0,0 +1,25 @@
+  failure_message do |context|
+    "Expected `#{error}` got `#{context.contract.errors.full_messages.join(",")}`"
+  end
+end
```
Yes I only started adding them, but definitely far from finished or exhaustive, feel free to play with the idea and add more :+1:

jjaffeux commented on plugins/chat/app/services/base.rb:
```diff
@@ -0,0 +1,186 @@
+        end
+      end
+
+      # @!method guardian(key, *args)
```
Yep it's related to an error I mentioned yesterday, I need to fix it for doc to update :p

jjaffeux commented:
> Minor nitpick, for some reason https://discourse.github.io/discourse/chat/backend/index.html has things like registerChatComposerButton, shouldn't that be in the Frontend section?

It's existing doc that I need to move around now that we have pattern for both frontend and backend

jjaffeux commented:
> @jjaffeux I also just added an `UpdateChannel` service in this commit [24691b3](https://github.com/discourse/discourse/commit/24691b3cfc8a1aec04e543f749c228075ecfffc7) . It was quite easy, though I did come across a new problem which we needed to consider. That is the issue where, for controllers at least, not all the parameters can be passed in, in the case of the Update endpoint you can pass in the channel `:name` but not the `:description`, and we don't want to change `:description` to null just from the absence of a parameter.

Thanks for try this with a different service, exactly what I wanted to see!

I removed the `change_attributes` addition as the general idea is to have "context" be the source of trust, it should move from top to bottom of the service. I had to make a few fixes to base service but you can now correctly do what you wanted in a `before_contract` block where `context` should be available an can be tweaked before going through the actual contract validation.

jjaffeux commented:
> The other issue is rate limiting -- for the channels controller, based on the value of a parameter we perform `auto_join_limiter`. Do we want to make it so the controller can easily access the resolved service object contract data? Or have a way to allow services to handle rate limits (not sure if this is a good idea since it's mostly a controller thing ATM)? Or something else, maybe just leave it how it is in the controller? Not sure.

Yes that's a tricky one. Some kind of chicken/egg problem. We have some logic in our `before_contract` which will shape the value of `auto_join_users` and we want to intercept here before the `service` is called.

I feel like this could be in the service, I agree it's mostly used in controllers, however it means we have to duplicate all this rate limiting logic each time it's used, means someone could use it in another controller without rate limiting it.

So far I would tend for a `before_service` block calling a `context.fail!`.

thoughts @Flink ?

jjaffeux commented:
@martin-brennan @Flink I would want this to be our main focus for the start of the week. Would want this merged before we start threading work.

martin-brennan commented on plugins/chat/spec/requests/api/chat_channels_controller_spec.rb:
```diff
@@ -421,7 +392,7 @@
       it "nullifies the field and doesn’t store an empty string" do
```
The description for this and the name one are not right now. The way the `UpdateChannel` service is set up now with `before_context` doesn't allow us to blank out the name or description (name is much less likely to be needed, but description is legit).

martin-brennan commented on plugins/chat/app/services/update_channel.rb:
```diff
@@ -0,0 +1,91 @@
+      before_contract { guardian(:can_edit_chat_channel?) }
+
+      before_contract do
+        context.name = (context.name || context.channel.name).presence
```
This is nicer, though I think still not quite right, since there is no way to even blank out name or description if necessary (though it's much less likely to want to do this for name). This is where the contortions I introduced helped, since you could inspect the original params passed in and see if they were even present. If description was present + nil, then we could blank it out, otherwise if it was not present we would not touch it.

Not sure how to express this without having the original params...

jjaffeux replied:
Yes this is all messed up ATM, feel free to fix if you want, I noted to come back to this tomorrow.

martin-brennan replied:
@jjaffeux I made this commit https://github.com/discourse/discourse/pull/19814/commits/ecf57a1062e68884d53b8136dcc3216dc11da595 to address this, I made the `@initial_context` hash available to the service so it can be inspected in `before_contract`.

martin-brennan commented on plugins/chat/app/controllers/api/chat_channels_status_controller.rb:
```diff
@@ -2,17 +2,17 @@
+    if result.success?
+      render_serialized(channel_from_params, ChatChannelSerializer, root: "channel")
+    else
+      # FIXME: implement failure handling
```
Why not use the handler I introduced, `handle_service_result` ?

jjaffeux replied:
mostly because I know @Flink is working on something too, so I didn't bother for now

martin-brennan reviewed (commented):
> Couldn't remember why a nil name or description attribute should stay nil and not default to channel name/description, didn't check, maybe we just rely on this all the time after instead of storing in the DB. Probably minor, will think of this later...

So from what I can tell, and I commented this, in some places it is legit to clear out a field (e.g. if you want to remove description from a channel). Thinking outside chat too, some fields on the user profile for instance should be clearable. Not sure the nicest way to express this without having access to the original `params`.

> I feel like this could be in the service, I agree it's mostly used in controllers, however it means we have to duplicate all this rate limiting logic each time it's used, means someone could use it in another controller without rate limiting it.

> So far I would tend for a before_service block calling a context.fail!.

For rate limiting, I do think this is a good idea, though I do wonder whether we need some sort of escape hatch for the case of Job -> Service or Service -> Service. We may also want to add this rate limiting failure as another top-level failure, like contract_failed or guardian_failed.

Keen to see what @flink thinks too.

> I would want this to be our main focus for the start of the week. Would want this merged before we start threading work. This is why I have removed ChatMessageCreatorV2 for now, as I think this is too complex to tackle in this PR.

100% agree with you, I don't want to start threading (with the exception of something like adding the site setting) without getting this PR merged. Hopefully we can merge by around Wednesday.

> Yes I only started adding them, but definitely far from finished or exhaustive, feel free to play with the idea and add more +1

Will add the guardian spec failure check.

martin-brennan commented on plugins/chat/app/services/update_channel.rb:
```diff
@@ -0,0 +1,91 @@
+      end
+
+      def publish_channel_update
+        # FIXME: this should become a dedicated service
```
Why have a whole service just for this, when the publisher already does this? It's just one line.

jjaffeux replied:
Yes Im unsure yet, two things which are bothering me in this state:
- it makes for services with different APIs/naming
- we need to think about errors happening inside this

martin-brennan commented on plugins/chat/app/services/update_channel.rb:
```diff
@@ -0,0 +1,91 @@
+      end
+
+      def auto_join_users_if_needed
+        # FIXME: this should become a dedicated service
```
Also not sure this needs to be its own service...maybe if I squint. What would it be called? `Service::AutoJoinChannel` or something?

martin-brennan commented on lib/tasks/documentation.rake:
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+task "documentation" do
```
Oh I just realised this is in core, shouldn't it be in `plugins/chat/lib/tasks/documentation.rake` ? If that's not possible fine, just wondering

jjaffeux replied:
The idea is to provide a general solution for plugins and core to generate documentation

martin-brennan replied:
No probs :+1:

Flink commented:
@jjaffeux @martin-brennan Just pushed some unfinished code regarding the helper to be used in controllers. I wrote a (almost) working example for `TrashChannel` in the channel controller. The code doesn’t break the existing code we have for the service.
I’ll make improvements tomorrow :slightly_smiling_face:

jjaffeux commented:
> @jjaffeux @martin-brennan Just pushed some unfinished code regarding the helper to be used in controllers. I wrote a (almost) working example for `TrashChannel` in the channel controller. The code doesn’t break the existing code we have for the service. I’ll make improvements tomorrow 🙂

Thanks, couldn't work on it after sadly :/ popper stuff took me too much time.

martin-brennan reviewed (commented):
Thanks Loic, I appreciate what you have added here and the pattern that is used in the controller, but I also feel as if `Chat::Endpoint` has a lot of complexity that we don't really need at the moment. Compare with https://github.com/discourse/discourse/pull/19814/files#diff-308c708040e8e1bd2b581d0053c653ad6c46e9b2ab1ac9866ff6bd62168263f9R40 `wrap_service` for example:

* We raise `Discourse::InvalidAccess` for guardian failures, which are already handled in `ApplicationController`, there is no specific message for these, so there is no need to handle each different guardian check failure differently
* We yield the contract failed messages with a 400 status, since most of the time we want to just respond with that, but the caller still has flexibility to do something else
* We yield success with the result, so the user can do whatever they want when it's successful

It's simple to understand and to use as the caller and doesn't involve any magic. Sorry, I am not trying to be a negative Nancy, but I really think we should go with simplicity for the controller handler and expand it over time when necessary, personally I think there is too much complexity and indirection in `Chat::Endpoint` . The usage in the controller does look nice, though I am not a fan of having to much with the `params` object, and `on_failed_policy` is just unnecessary when we can just raise `InvalidAccess`.

Interested to see what @jjaffeux thinks as well.

martin-brennan commented:
@jjaffeux this change https://github.com/discourse/discourse/commit/45c3c0b19099fec5627d9649892451d376418a9d to Collection errors locally at this line, saying `module` is not defined:

```javascript
module.exports = {
  default: Collection,
};
```

I changed it back to `export default class Collection` for now.

martin-brennan commented:
Something added here is adding tons of these spec errors, this does not happen on main (included an example spec):

```
TopicsController crawler when a crawler with canonical_url set to topic.url when embed_set_canonical_url is false
        Failure/Error: opts = context.eval(buffer)

        MiniRacer::RuntimeError:
          Error: Parser rule not found: fragments_join
        # JavaScript at Ruler.before (<anonymous>:3455:13)
        # JavaScript at Function.<anonymous> (<anonymous>:132:24)
        # JavaScript at MarkdownIt.use (<anonymous>:8289:12)
        # JavaScript at <anonymous>:347:21
        # JavaScript at Array.forEach (<anonymous>)
        # JavaScript at setupMarkdownEngine (<anonymous>:338:26)
        # JavaScript at setup (<anonymous>:482:5)
        # JavaScript at buildOptions (<anonymous>:93:36)
        # JavaScript at <anonymous>:27:17
        # ./vendor/bundle/ruby/3.1.0/gems/mini_racer-0.6.3/lib/mini_racer.rb:228:in `eval_unsafe'
        # ./vendor/bundle/ruby/3.1.0/gems/mini_racer-0.6.3/lib/mini_racer.rb:228:in `block (2 levels) in eval'
        # ./vendor/bundle/ruby/3.1.0/gems/mini_racer-0.6.3/lib/mini_racer.rb:369:in `timeout'
        # ./vendor/bundle/ruby/3.1.0/gems/mini_racer-0.6.3/lib/mini_racer.rb:227:in `block in eval'
        # ./vendor/bundle/ruby/3.1.0/gems/mini_racer-0.6.3/lib/mini_racer.rb:225:in `synchronize'
        # ./vendor/bundle/ruby/3.1.0/gems/mini_racer-0.6.3/lib/mini_racer.rb:225:in `eval'
```

Here is where that rule is used:

https://github.com/discourse/discourse/blob/f31f0b70f82c43d93220ce6fc0d4f57440452f37/app/assets/javascripts/pretty-text/engines/discourse-markdown/bbcode-inline.js#L167-L170

Not sure what has caused this, looking...

martin-brennan commented:
Found why `Error: Parser rule not found: fragments_join` was happening.

It's the addition of `"tidy-jsdoc": "^1.4.1"` in `app/assets/discourse/package.json`, it adds `jsdoc` which in turn relies on an older version of markdown-it than we use in core:

```
markdown-it "^12.3.2"
markdown-it-anchor "^8.4.1"
```

And our version:

https://github.com/discourse/discourse/blob/cd6852e0a06263b4c3f2b82720e959dfdb402475/app/assets/javascripts/discourse/package.json#L83-L83

From what I can tell this is just a JSDoc theme, so removing this in https://github.com/discourse/discourse/pull/19814/commits/69a783ea630afdc9444816ab9cb28c5b1072c8e7 so everything works again.

Flink commented:
> Sorry, I am not trying to be a negative Nancy, but I really think we should go with simplicity for the controller handler and expand it over time when necessary, personally I think there is too much complexity and indirection in `Chat::Endpoint` . The usage in the controller does look nice, though I am not a fan of having to much with the `params` object, and `on_failed_policy` is just unnecessary when we can just raise `InvalidAccess`.

No need to be sorry, it’s a good thing to raise any concern you might have :wink:
As stated yesterday, the code isn’t finished yet and while it may seem complex at first glance, it’s not that complex, it’s just a way of handling any possible outcome always in the same way. The wrapper you made does the job, but I’m thinking long term when we want to handle more complex cases. It’s easy to add other cases if it works for us, we just have to match a convention decided by us in the result object. As you said yourself, the usage in the controller is nice and that’s my goal. As we can just express very clearly what happens in which situation, anybody looking at the controller code will grasp what’s going on at a glance. What I like is that it’s very straightforward, you don’t need to perform checks or whatever, you just describe what will happen for a given result. I also know for a fact that this pattern works as I’ve used it in a project where Trailblazer was used (so yes it’s _heavily_ inspired by it :sweat_smile:).

I’ll make the improvements we discussed yesterday with @jjaffeux, hopefully they’ll be ready by the end of my day and maybe some of your concerns will vanish :slightly_smiling_face:

Flink commented on plugins/chat/app/controllers/api_controller.rb:
```diff
@@ -4,10 +4,68 @@ class Chat::Api < Chat::ChatBaseController
   before_action :ensure_logged_in
   before_action :ensure_can_chat
 
+  def result
+    @_result
+  end
+
```
Yes, I’ve borrowed it from trailblazer :sweat_smile: 
The `Chat::Endpoint` class will set automatically the result object from the service in this instance variable and we make it accessible through this getter.

Flink replied:
Actually that changed a bit with my latest changes. Now that’s the `run_service` helper in `Chat::Api` that is responsible for this. It will be easier to maintain/change/whatever.

Flink commented:
Ok that’s it for my day, so I think the `with_service` helper and the `Endpoint` class are more or less done.
I’ve made some modifications to the service base class: now we can define policies with the `policy` keyword. The goal is to put our guardians in it but it can contain any code really. We can define more than one policy and if that’s the case, they’ll run in their definition order. When a policy returns a falsy value, it will fail and a `result.policy.name_of_the_policy` key will be added to the result object. This is also a result object so we can query it like this: `result[:"result.policy.name"].failure?`.

The `with_service` helper now also comes with default actions by default, meaning that typically we’ll just have to run a service like this: `with_service(MyService)`. There’s a new `extra_params` parameter to provide options without having to change `params` manually. The `TrashChannel` service and its endpoint have been rewritten to take advantage of these new features. 

I also added some dedicated matchers and some specs for the endpoint class. I’ve tried to be backward compatible and not break your work @martin-brennan while implementing all this :sweat_smile: (I’ve updated some services and helpers).
There are still some issues though that need to be addressed, like what to do if for whatever reason we don’t have an object we need but, we have a policy before a contract. I see various solutions to this, like having an action to fetch models inside the service or being able to execute actions as they’re defined (so we could have a policy, a contract, another policy, etc.)

martin-brennan commented:
@Flink thank you for making further improvements. Now that I see the complete picture, it hits me differently today, and I quite like the API you have added and the defaults. I am still not a huge fan of the `Chat::Endpoint` class generally, it "feels" complex to me in one of those wishy-washy bad-aura kind of ways, but that's my problem. I figure we likely will not have to touch it very often, and as long as it's well-documented and tested it's fine by me. I changed the other API endpoints to use `with_service` and deleted my own helpers, will leave it to @jjaffeux for the final look :+1:

jjaffeux commented:
@martin-brennan Wanted to update you on this. We have been working with @Flink today to migrate another service to this: https://github.com/discourse/discourse/pull/19814/commits/57eb3a2e1614009060691ad9269f6fae180c9d64

My goal is to have this merged by Wednesday (my morning).

The big thing I want to finish tomorrow with @Flink is better developer experience while coding a service, our goal is to have an output similar to this:

```
expected service result 'result.policy.invalid_access' does not exist

step 1: [model] foo
step 2: [policy] invalid_access
step 3: baz result.membership  => "..."
```

martin-brennan commented:
Thanks for the update @jjaffeux keen to see how it turns out. Things have changed a lot here with DSL, steps, policies etc. so would be good to have a high level overview at the end, we will need one for other engineers coming on to the project too. LMK if there is anything you need from me, but it seems like you and Loic have this under control now :+1:

ZogStriP commented on plugins/chat/spec/services/update_channel_spec.rb:
```diff
@@ -0,0 +1,107 @@
+  context "when channel is a category one" do
+    context "when a valid user provides valid params" do
+      let(:message) do
+        MessageBus.track_publish(ChatPublisher::CHANNEL_EDITS_MESSAGE_BUS_CHANNEL) { result }.first
```
Not to be addressed in that PR, but we've had some flaky test whenever two pieces of code would publish a message and would break the specs that are looking for the `first` message in the bus. Do/should we have a way to filter for a certain type of message and then do `.first` on it? 🤔

jjaffeux replied:
yes agreed, generally I think `include?` on a map of events is just enough
